File: 082.Remove Duplicates from Sorted List II.java
/*
    Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

    For example,
    Given 1->2->3->3->4->4->5, return 1->2->5.
    Given 1->1->1->2->3, return 2->3.
     */

    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */

    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode one = dummy;
        ListNode two = dummy.next;
        int now = head.val;
        while (two != null) {
            while(two.next != null && two.val == two.next.val){
                two = two.next;
            }
            if(one.next == two){
                one = one.next;
            } else {
                one.next = two.next;
            }
            two = two.next;
        }
        return dummy.next;
    }

File: LeetCode/082.Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        prenode = dummy = ListNode(0)
        dummy.next = head
        ptr = head
        is_dup = False
        while ptr.next != None:
            if ptr.val == ptr.next.val:
                ptr.next = ptr.next.next
                is_dup = True
            else:
                if is_dup:
                    prenode.next = ptr.next
                else:
                    prenode = ptr
                ptr = ptr.next
                is_dup = False
        if is_dup:
            prenode.next = None
        return dummy.next
